core/blockchain.go

```diff
@@ -942,6 +942,27 @@ func SetReceiptsData(config *params.ChainConfig, block *types.Block, receipts ty
 			logIndex++
 		}
 	}
+	// in a block, txns are processed before staking txns
+	for j := len(transactions); j < len(transactions)+len(stakingTransactions); j++ {
+		// The transaction hash can be retrieved from the staking transaction itself
+		receipts[j].TxHash = stakingTransactions[j].Hash()
+
+		// The used gas can be calculated based on previous receipts
+		if j == 0 {
```

Comment:
We can have code gen completely avoid this branch (huge free performance win) since you are only needing it once. Do this j == 0 assignment before you loop adjust the for loop's indexing as needed

Reply:
@fxfactorial wait this code looks weird. shouldn't this me `if j = len(transactions)` ? cc: @rlan35

Reply:
done. good point.